httpx/middleware/query_normalizer.go

```diff
@@ -0,0 +1,46 @@
+package middleware
+
+import (
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+// WithQueryNormalizer fixes wrong php query string handling as array
+func WithQueryNormalizer(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if i := strings.Index(r.URL.RawQuery, "["); i == -1 {
+			return
+		}
+
+		q := r.URL.Query()
+		normalizedQuery := make(url.Values, 0)
```

Comment:
pessimistically assign the map size

Reply:
You could use `len(q)`, but for small maps it would make very little to no difference. I would leave it up to @rafalmnich .

Reply:
Not changing anything in benchmarks with even for query with many changes like:
/?param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val&param[smth][0]=zero&param[smth][0][val][1][asdf][4][asetrhyty][5][ewrytu][7][aerterwert][7]=val